### PR TITLE
router: support for x-forwarded-host header.

### PR DESCRIPTION
**Pull request by arun-vasudev**
Commit Message: _Support for x-forwarded-host header._

Additional Description:
_If, for a request, the host/authority header is changed when the
request is proxied, set the x-forwarded-host header as:
x-forwarded-host = append(x-forwarded-host, host)_

Risk Level: _Low_

Testing: _unit test and manual testing_

Docs Changes:
_Added x-forwarded-host header in HTTP header manipulation
(configuration/http/http_conn_man/headers).
Under the host_rewrite options in route_components.proto
(in v2, v3, and v4alpha), added that when host is rewritten,
XFH header is appended with the original value of host header._

Release Notes: _router: added support for *x-forwarded-host* header._

Platform Specific Features: _N/A_

Fixes https://github.com/envoyproxy/envoy/issues/5940

### CODE REVIEW

**Review by lizan — CHANGES_REQUESTED** [review]
API LGTM, do we need a feature guard for this? cc @alyssawilk

Please also add a release note.

**Comment by arun-vasudev** [comment]
Hi @lizan, thanks for the review. Just one thing, though. I had added notes in docs/root/version_history/current.rst. Is there a different 'release notes'?

**Comment by lizan** [comment]
> Hi @lizan, thanks for the review. Just one thing, though. I had added notes in docs/root/version_history/current.rst. Is there a different 'release notes'?

That's enough, thank you. Please don't force push since it screw up review history on GitHub. Defer to @alyssawilk for the rest of review.

**Comment by arun-vasudev** [comment]
Sure @lizan. This time, the force push only took care of a rebase merge conflict. So, there have not been any changes since your review. But sure. I'll keep that in mind.

File: api/envoy/config/route/v3/route_components.proto
Comment (alyssawilk) [inline api/envoy/config/route/v3/route_components.proto:906]: I'm inclined to wonder if which way this should default. It seems no one has needed it so far, so making it optional and default-off seems both in line with our new API rules (cc @markdroth) and save bytes on the wire.

If we opt to leave it as is, we definitely need to runtime guard (as mentioned in CONTRIBUTING.md) but I'd hold off on guarding until we settle on config.

File: source/common/http/utility.h
Comment (alyssawilk) [inline source/common/http/utility.h:191]: have you checked all call sites for setHost?  I see at least one other in extensions which looks like it might need to be updated.

File: tools/spelling/spelling_dictionary.txt
Comment (alyssawilk) [inline tools/spelling/spelling_dictionary.txt:361]: optional (since we do this for XFF) but it might be worth just spelling this one out at the 3 places we mention XFH since it's not as well known as XFF.

**Review by alyssawilk — COMMENTED** [review]
This looks great overall - thanks for helping to make Envoy more standards compliant here :-)

File: source/common/http/utility.h
Comment (yanavlasov) [reply source/common/http/utility.h:191]: I think this comes down to whether we want to make just part of router host rewrite determined by router config or a global behavior that applies to filter that rewrite host for some reason. If we make this behavior dependent on the router config then extensions need to either add their own config options that control XFH or be modified to always do it. If we go with this change I would rather make it in separate PRs and scope this one just to the router filter.

I think I have a preference for scoping this PR to the router hots rewrite only and not extend it to the filters.

**Review by yanavlasov — CHANGES_REQUESTED** [review]
/wait

**Pull request closed without merging**
